dask/dataframe/multi.py

```diff
@@ -419,6 +420,239 @@ def merge(left, right, how='inner', on=None, left_on=None, right_on=None,
+        if i1 > i:
+            result.append(J)
+            J = []
+        i, j = i1, j1
+
+    return result
+
+
+def merge_asof_padded(left, right, prev=None, next=None, **kwargs):
+    """ merge_asof but potentially adding rows to the beginning/end of right """
+    if left is None or right is None:
+        raise ValueError('dafuq did u do')
+
+    frames = []
+    if prev is not None:
+        frames.append(prev)
+    frames.append(right)
+    if next is not None:
+        frames.append(next)
+
+    return pd.merge_asof(left, pd.concat(frames), **kwargs)
+
+
+def merge_asof_indexed(left, right, **kwargs):
+    left = fix_overlap(left)
+    right = fix_overlap(right)
+
+    dsk = dict()
+    name = 'asof-join-indexed-' + tokenize(left, right, **kwargs)
+    meta = pd.merge_asof(left._meta_nonempty, right._meta_nonempty, **kwargs)
+
+    dependencies = [left, right]
+    tails = heads = None
+    if kwargs['direction'] in ['backward', 'nearest']:
+        tails = compute_tails(right, by=kwargs['right_by'])
+        dependencies.append(tails)
+    if kwargs['direction'] in ['forward', 'nearest']:
+        heads = compute_heads(right, by=kwargs['right_by'])
+        dependencies.append(heads)
+
+    for i, J in enumerate(pair_partitions(left.divisions, right.divisions)):
+        frames = []
+        for j, lower, upper in J:
+            slice = (methods.boundary_slice, (left._name, i),
+                     lower, upper, False)
+            tail = (tails._name, j) if tails is not None else None
+            head = (heads._name, j) if heads is not None else None
+            frames.append((apply, merge_asof_padded, [slice, (right._name, j),
+                           tail, head], kwargs))
+        dsk[(name, i)] = (pd.concat, frames)
+
+    graph = HighLevelGraph.from_collections(name, dsk,
+                                            dependencies=dependencies)
+    return new_dd_object(graph, name, meta, left.divisions)
+
+
+@wraps(pd.merge_asof)
+def merge_asof(left, right, on=None, left_on=None, right_on=None,
+               left_index=False, right_index=False, by=None, left_by=None,
+               right_by=None, suffixes=('_x', '_y'), tolerance=None,
+               allow_exact_matches=True, direction='backward'):
+    if direction not in ['backward', 'forward', 'nearest']:
+        raise ValueError("Invalid merge_asof direction. Choose from 'backward',\
+              'forward', or 'nearest'")
+
+    kwargs = {'on': on, 'left_on': left_on, 'right_on': right_on,
+              'left_index': left_index, 'right_index': right_index,
+              'by': by, 'left_by': left_by, 'right_by': right_by,
+              'suffixes': suffixes, 'tolerance': tolerance,
+              'allow_exact_matches': allow_exact_matches,
+              'direction': direction}
+
+    if left is None or right is None:
+        raise ValueError("Cannot merge_asof on empty DataFrames")
+
+    if isinstance(left, pd.DataFrame) and isinstance(right, pd.DataFrame):
```

Comment:
@mrocklin what's best for cuDF here? Would a dispatchable `merge_asof`, and replace these checks with `is_dataframe_like` be best?

Reply:
I think that we should use `is_dataframe_like` universally.

I think that we should continue to use `pd.merge_asof` until some other pandas implementation supports the operation, at that point I think that we would either try to add a method, or a dispatchable function.  I wouldn't worry about this until that happens though.

Reply:
So, to be clear @codercody , if you can replace checks like 

```python
isinstance(df, pd.DataFrame)
```

with 

```python
is_dataframe_like(df)
```

That would be welcome.

Reply:
I was trying to copy line 346 in the `merge` implementation, where they separately handled when `left` and `right` are actually `pd.DataFrame`s and not `dd.DataFrame`s. If you don't want me to handle this case, then I'm just going to delete this check altogether. Is that fine?